pkg/provisioners/message.go

```diff
@@ -62,3 +72,58 @@ func headerSet(headers []string) map[string]bool {
 	}
 	return set
 }
+
+// History returns the list of hosts where the message has been into
+func (m *Message) History() []string {
+	if m.Headers == nil {
+		return nil
+	}
+	if h, ok := m.Headers[MessageHistoryHeader]; ok {
+		return decodeMessageHistory(h)
+	}
+	return nil
+}
+
+// AppendToHistory append a new host at the end of the list of hosts of the message history
+func (m *Message) AppendToHistory(host string) {
+	host = cleanupMessageHistoryItem(host)
+	if host == "" {
+		return
+	}
+	history := m.History()
```

Comment:
Why not have `History()` return a len-zero list of history, rather than `nil`? This would probably make it easier to work with in general.

Reply:
This could would become:
```golang
host = cleanup(host)
if host == "" {
  return
}
m.SetHistory(m.History(), host)
```